### PR TITLE
Import reference

### PR DESCRIPTION
**Pull request by cgwalters**
Import Reference type from rust-oci-client

Almost all tools which operate with OCI will end up
wanting to parse and use proper references. Import
the code from the rust-oci-client crate.

I imported the code with history with this blog
which I found helpful:
https://savorywatt.com/2015/01/25/move-files-and-folders-between-git-repos-using-patches/

Changes:

- Move `regex.rs` code inline into this file
- Switch to std `LazyLock` instead of `lazy_static` crate

For more see:

- https://github.com/containers/oci-spec-rs/issues/205
- https://github.com/oras-project/rust-oci-client/issues/159

Signed-off-by: Colin Walters <walters@verbum.org>

### CODE REVIEW

File: src/distribution/reference.rs
Comment (cgwalters) [inline src/distribution/reference.rs:270]: This could also clearly use https://github.com/containers/oci-spec-rs/commit/5601907d5c4ead6ae42ea191f485726fd6179f98 etc. but probably best to keep code changes minimal until we've deduplicated.

**Comment by flavio** [comment]
I'm totally in favor of this PR. Once merged I'll update `rust-oci-client` to consume `oci-spec-rs`

File: src/distribution/reference.rs
Comment (saschagrunert) [inline src/distribution/reference.rs:180]: I guess we could use `getset` for those.

**Review by saschagrunert — APPROVED** [review]
@utam0k PTAL. I just have a non-blocking nit.

File: src/distribution/reference.rs
Comment (thomastaylor312) [reply src/distribution/reference.rs:270]: Yep I agree here. We just had this as a string for simplicity, but it would probably be better to have it be a concrete digest type. Definitely going to be easier to wait until this is merge though

**Review by thomastaylor312 — APPROVED** [review]
Thanks for pulling this all in! Looking forward to using it downstream in oci-client

**Comment by cgwalters** [comment]
Reusing this PR as a communication channel...I think it'd probably make sense to add some folks who maintained/wrote the reference type as reviewers on this repository right? @flavio will you be able to watch for changes on this repo too? Any others to add?

**Comment by flavio** [comment]
Fine with me. @thomastaylor312 could be interested too

**Comment by saschagrunert** [comment]
> @saschagrunert ok to add those two folks to have write/reviewer access here?

Yes! Having more maintainers would be awesome within this project.

> On a related topic I notice this repository (unlike others in the containers/ ecosystem) doesn't have branch protection rules set up which we really should have, i.e. require a PR to push to git main etc. Any objections to me also adding those?

Sure feel free to add it, thanks you for double ckecking!

**Comment by cgwalters** [comment]
Related to this...as this project has grown some, does anyone have thoughts on possibly submitting an application to the [CNCF](https://www.cncf.io/) ? I think it'd be a good home for a project such as this; we're a multi-vendor/organization ASL2.0 project.

(And if we do I'd like to add https://github.com/containers/ocidir-rs/ too)

**Comment by saschagrunert** [comment]
@cgwalters would that project fit more specifically into the [OCI](https://opencontainers.org/)?

**Comment by cgwalters** [comment]
Seems likely yep

**Pull request merged**
